### PR TITLE
Fastspeech

### PR DESCRIPTION
**Pull request by bloodraven66**
Uses durations acquired from CTC based forced alignment.

### CODE REVIEW

**Comment by mravanelli** [comment]
Hi @bloodraven66,
thank you for this PR! Unfortunately, I cannot run it because I don't have the durations. Here are my comments:
1. As for the duration, I would suggest to put them on the web (e.g., on dropbox). When we run download `prepare_ljspeech_durations_and_predefined_splits`, we can automatically download it and store in the data folder. We do something similar in other recipes (e.g., https://github.com/speechbrain/speechbrain/blob/develop/recipes/timers-and-such/prepare.py#L56). Once done that, you can remove "duration_path" in the yaml file.
2. We have to update the README file with the fastspeech recipe as well.
3. Tests are failing (see below). Note that you can run the tests locally as well simply by running `pytest` in the main speechbrain folder. You can also run test by typing `tests/.tests/.run-unittests.sh`  `tests/.run-doctests.sh` `tests/.run-linters.sh` , etc. 
4. Normally the output folder is not hard-coded (it should be `output_folder: !ref results/fastspeech/<seed>`  here)
5. The datafolder should be (`data_folder: !PLACEHOLDER # e.g., /data/Database/LJSpeech-1.1`)

@BenoitWang, could you please share your feedback as well? The feedback from @flexthink is welcome as well.

File: recipes/LJSpeech/TTS/fastspeech/train.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+
+def to_gpu(x):
+    x = x.contiguous()
+
+    if torch.cuda.is_available():
+        x = x.cuda(non_blocking=True)
+    return x
+
+
+def criterion(predictions, targets, log_scale_durations=True):
```
Comment (BenoitWang) [inline recipes/LJSpeech/TTS/fastspeech/train.py:318]: better to remove the loss func into Fastspeech.py and call it from the yaml file. And for improvement maybe we can try adding L1 for mel_loss too.

File: speechbrain/lobes/models/FastSpeech.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+def upsample(feats, durs, pace=1, padding_value=0.0):
+    """upsample encoder ouput according to durations
+    Arguments
+    ---------
+    feats: torch.tensor
+        batch of input tokens
+    durs: torch.tensor
+        durations to be used to upsample
+    pace: int
```
Comment (BenoitWang) [inline speechbrain/lobes/models/FastSpeech.py:296]: `pace` is an important parameter but not used in the upsample function, while changing the speed is one of the best features of Fastspeeh. I think we should use `int(pace * durs[i])` instead of `durs[i]`, and we add it as a hyperparameter in yaml

**Review by BenoitWang — COMMENTED** [review]
Hi, please check my comments here.

File: recipes/LJSpeech/TTS/ljspeech_prepare.py [hunk 3]
```diff
@@ -24,6 +25,81 @@
...
+        "wavs":wavs,
+        "seed": seed,
+    }
+
+    if not os.path.exists(save_folder):
+        os.makedirs(save_folder)
+
+    #check that metadata files exists in specified path
+    for filename in [train, valid, test, wavs, duration]:
+        assert os.path.exists(filename), f"{filename} not found"
```
[the remark this answers is not in the document]
Comment (bloodraven66) [reply recipes/LJSpeech/TTS/ljspeech_prepare.py:59]: I have removed the predefined splits and added on to prepare_ljspeech (works by default for tacotron)

File: recipes/LJSpeech/TTS/ljspeech_prepare.py [hunk 4]
```diff
@@ -24,6 +25,81 @@
...
+        for metadata_line in lines:
+            data.append({
+                    "ID": metadata_line[0],
+                    "speaker_id": "0",
+                    "wav": os.path.join(wavs, f"{metadata_line[0]}.wav"),
+                    "label": metadata_line[1],
+                    "durations": os.path.join(duration, metadata_line[0]+'.npy'),
+                    })
+
+        path = os.path.join(save_folder, name+'.csv')
```
[the remark this answers is not in the document]
Comment (bloodraven66) [reply recipes/LJSpeech/TTS/ljspeech_prepare.py:92]: changed

File: recipes/LJSpeech/TTS/fastspeech/train.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+        if self.last_batch is None:
+            return
+        tokens, input_lengths, *_ = self.last_batch
+        token = tokens[index][:input_lengths[index]]
+        mel_post, predicted_durations =  self.hparams.model(token.unsqueeze(0))
+        self.hparams.progress_sample_logger.remember(
+            infer_output=self.process_mel(mel_post, [len(mel_post[0])])
+        )
+
+    def run_vocoder(self):
```
[the remark this answers is not in the document]
Comment (bloodraven66) [reply recipes/LJSpeech/TTS/fastspeech/train.py:215]: is the train test split for vocoder same as tacotron split?

[the remark this answers is not in the document]
Comment (BenoitWang) [reply recipes/LJSpeech/TTS/fastspeech/train.py:215]: Yes they should be the same.

**Comment by anautsch** [comment]
Hi @bloodraven66 just ran the workflow to check status (even though you might still be developing/drafting here).

There were some errors/warnings, you should be able to reproduce them locally :)
(Please try to fetch the latest develop branch into your PR, too.)
* please enlist `recipes/LJSpeech/TTS/fastspeech/hparams/train.yaml` at `tests/recipes.csv`
* please run `pytest tests/consistency` and `./tests/.run-linters.sh`
